site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/impl/SiteManageGroupSectionRoleHandler.java

```diff
@@ -854,268 +870,376 @@ public String processCancelDelete()
+                M_log.warn(this + ".processAutoCreateGroup: cannot find site " + site.getId(), e);
+                return null;
+            } 
+            catch (PermissionException e) {
+                M_log.warn(this + ".processAutoCreateGroup: cannot find site " + site.getId(), e);
+                return null;
+            }
+        }
+
+        return "done";
+    }
+    
+    /**
+     * SAK-29373 - Create random groups of users from the given roster.
+     * This could be random number of users in specified number of groups,
+     * or vice versa.
+     * 
+     * @param rosters a list of provider IDs (should only ever be one in the list)
+     * @param unit the number of groups or users per group requested
+     */
+    private void createRandomGroupsForRoster( List<String> providerIDs, int unit )
+    {
+        String groupTitle = rosterGroupSplit ? rosterGroupTitleGroup: rosterGroupTitleUser;
+        if( !StringUtils.isEmpty( groupTitle ) )
+        {
+            List<String> userList = new ArrayList<>();
+            for( String providerID : providerIDs )
+            {
+                try
+                {
+                    Set<Membership> sectionMembers = cms.getSectionMemberships( providerID );
+                    for( Membership member : sectionMembers )
+                    {
+                        String userEID = member.getUserId();
+                        try
+                        {
+                            String userID = userDirectoryService.getUserId( userEID );
+                            if( !userList.contains( userID ) )
+                            {
+                                userList.add( userID );
+                            }
+                        }
+                        catch( UserNotDefinedException ex )
+                        {
+                            M_log.debug( this + ".createRandomGroupsForRoster: user not defined = " + userEID, ex );
+                        }
+                    }
+                }
+                catch( IdNotFoundException ex )
+                {
+                    M_log.debug( this + ".createRandomGroupsForRoster: can't find section for provider ID = " + providerID, ex );
+                }
+            }
+
+            createRandomGroups( rosterGroupSplit, userList, groupTitle, unit );
+        }
+    }
+
+    /**
+     * Common algorithm called to create random groups.
+     * @param isGroupSplit
+     * @param userIDs
+     * @param groupTitle
+     * @param unit 
+     */
+    private void createRandomGroups( boolean isGroupSplit, List<String> userIDs, String groupTitle, int unit )
+    {
+        int numOfGroups = -1;
+        int numOfUsersPerGroup = -1;
+        int totalNumOfUsers = userIDs.size();
+        
+        // Determine number of users per group and number of groups to create
+        if( isGroupSplit )
+        {
+            numOfGroups = unit > totalNumOfUsers ? totalNumOfUsers : unit;
```

Comment:
Remainders will be addressed in a future ticket. We have plans to provide another option in the future to determine how the user wishes to handle remainders (either add the remainders to a new group, or add them randomly to groups within the set being created).